File: Difficulty: Medium/K Closest Points to Origin/k-closest-points-to-origin.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<pair<int, vector<int>>> q;
        
        for (vector<int>& point : points) {
            int x = point[0], y = point[1];
            int sqDis = x * x + y * y;

            q.push({sqDis, point});
            if (q.size() > k) {
                q.pop();
            }
        }

        vector<vector<int>> ans;
        while (!q.empty()) {
            ans.push_back(q.top().second);
            q.pop();
        }

        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int k;
        cin >> k;
        int n;
        cin >> n;
        vector<vector<int>> points(n, vector<int>(2));
        for (int i = 0; i < n; i++) {
            cin >> points[i][0] >> points[i][1];
        }
        Solution ob;
        vector<vector<int>> ans = ob.kClosest(points, k);
        sort(ans.begin(), ans.end());
        for (const vector<int>& point : ans) {
            cout << point[0] << " " << point[1] << endl;
        }
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/K Closest Points to Origin/k-closest-points-to-origin.java
class Pair {
    double dist;
    int point[];
    Pair(double dist,int point[]){
        this.dist = dist;
        this.point = new int[point.length];
        for(int i = 0;i<point.length;i++)
            this.point[i]=point[i];
    }
}

class Solution {
    public ArrayList<ArrayList<Integer>> kClosest(int[][] points, int k) {
        PriorityQueue<Pair> minHeap= new PriorityQueue<>((a,b) -> Double.compare(a.dist,b.dist));
        for(int i[]:points) {
            double eDist = (double)i[0] * i[0] + i[1] * i[1];
            minHeap.add(new Pair(Math.sqrt(eDist),i));
        }
        
        ArrayList<ArrayList<Integer>> ans = new ArrayList<>();
        while(k-- != 0){
            ArrayList<Integer> temp = new ArrayList<>();
            Pair p = minHeap.poll();
            if(p.point.length == 2){
                temp.add(p.point[0]);
                temp.add(p.point[1]);
            }
            ans.add(temp);
        }
        return ans;
    }
}